lamindb/curators/__init__.py

```diff
@@ -703,22 +730,186 @@ def save_artifact(
+
+    Args:
+        dataset: The SpatialData-like object to validate & annotate.
+        schema: A `Schema` object that defines the validation constraints.
+
+    Example::
+
+        import lamindb as ln
+        import bionty as bt
+
+        # define sample schema
+        sample_schema = ln.Schema(
+            name="blobs_sample_level_metadata",
+            features=[
+                ln.Feature(name="assay", dtype=bt.ExperimentalFactor).save(),
+                ln.Feature(name="disease", dtype=bt.Disease).save(),
+                ln.Feature(name="development_stage", dtype=bt.DevelopmentalStage).save(),
+            ],
+            coerce_dtype=True
+        ).save()
+
+        # define table obs schema
+        blobs_obs_schema = ln.Schema(
+            name="blobs_obs_level_metadata",
+            features=[
+                ln.Feature(name="sample_region", dtype="cat[ULabel]").save(),
+            ],
+            coerce_dtype=True
+        ).save()
+
+        # define table var schema
+        blobs_var_schema = ln.Schema(
+            name="blobs_var_schema",
+            itype=bt.Gene.ensembl_gene_id,
+            dtype=int
+        ).save()
+
+        # define composite schema
+        spatialdata_schema = ln.Schema(
+            name="blobs_spatialdata_schema",
+            otype="SpatialData",
+            components={
+                "sample": sample_schema,
+                "table:obs": blobs_obs_schema,
+                "table:var": blobs_var_schema,
+        }).save()
+
+        # curate a SpatialData
+        spatialdata = ln.core.datasets.spatialdata_blobs()
+        curator = ln.curators.SpatialDataCurator(spatialdata, spatialdata_schema)
+        try:
+            curator.validate()
+        except ln.errors.ValidationError as error:
+            print(error)
+
+        # add new categorical values and standardize
+        curator.slots["sample"].cat.add_new_from(key="developmental_stage")
+        curator.slots["sample"].cat.standardize(key="disease")
+        curator.slots["table:obs"].cat.add_new_from(key="sample_region")
+
+        # validate again (must pass now) and save artifact
+        artifact = curator.save_artifact(key="example_datasets/spatialdata1.zarr")
+        assert artifact.schema == spatialdata_schema
+    """
+
+    def __init__(
+        self,
+        dataset: SpatialData | Artifact,
+        schema: Schema,
+        *,
+        sample_metadata_key: str | None = "sample",
+    ) -> None:
+        super().__init__(dataset=dataset, schema=schema)
+        if not data_is_spatialdata(self._dataset):
+            raise InvalidArgument("dataset must be SpatialData-like.")
+        if schema.otype != "SpatialData":
+            raise InvalidArgument("Schema otype must be 'SpatialData'.")
+
+        # in form of {table_key: var_field}
+        self._var_fields: dict[str, FieldAttr] = {}
+        # in form of {table_key: categoricals}
+        self._categoricals: dict[str, dict[str, FieldAttr]] = {}
```

Comment:
How about move `self._var_fields` and `self._categoricals` initialization into SlotsCurator?

Reply:
The `AnnDataCurator` isn't using these. They are internal variables but I'd probably still keep them only in MuData and SpatialDataCurator. What do you think?